File: src/main/java/com/uiptv/db/ChannelDb.java
package com.uiptv.db;

import com.uiptv.model.Account;
import com.uiptv.model.Category;
import com.uiptv.model.Channel;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import static com.uiptv.db.DatabaseUtils.DbTable.CATEGORY_TABLE;
import static com.uiptv.db.DatabaseUtils.DbTable.CHANNEL_TABLE;
import static com.uiptv.db.DatabaseUtils.insertTableSql;
import static com.uiptv.db.DatabaseUtils.validatedTableName;
import static com.uiptv.db.SQLConnection.connect;

public class ChannelDb extends BaseDb {
    private static ChannelDb instance;
    private static final int BATCH_SIZE = 1000;


    public ChannelDb() {
        super(CHANNEL_TABLE);
    }

    public static synchronized ChannelDb get() {
        if (instance == null) {
            instance = new ChannelDb();
        }
        return instance;
    }

    public static void insert(Channel channel, Category category) {
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(insertTableSql(CHANNEL_TABLE))) {
            statement.setString(1, channel.getChannelId());
            statement.setString(2, category.getDbId());
            statement.setString(3, channel.getName());
            statement.setString(4, channel.getNumber());
            statement.setString(5, channel.getCmd());
            statement.setString(6, channel.getCmd_1());
            statement.setString(7, channel.getCmd_2());
            statement.setString(8, channel.getCmd_3());
            statement.setString(9, channel.getLogo());
            statement.setInt(10, channel.getCensored());
            statement.setInt(11, channel.getStatus());
            statement.setInt(12, channel.getHd());
            statement.setString(13, channel.getDrmType());
            statement.setString(14, channel.getDrmLicenseUrl());
            statement.setString(15, channel.getClearKeysJson());
            statement.setString(16, channel.getInputstreamaddon());
            statement.setString(17, channel.getManifestType());
            statement.execute();
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to execute query", e);
        }
    }

    private static void deleteAll(String categoryId) {
        String sql = "DELETE FROM " + validatedTableName(CHANNEL_TABLE) + " WHERE categoryId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, categoryId);
            statement.execute();
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to execute delete all query", e);
        }
    }

    public List<Channel> getChannels(String dbId) {
        return getAll(" WHERE categoryId=?", new String[]{dbId});
    }

    public int getChannelCountForAccount(String accountId) {
        String sql = "SELECT COUNT(*) FROM " + validatedTableName(CHANNEL_TABLE) +
                " WHERE categoryId IN (" +
                "SELECT id FROM " + validatedTableName(CATEGORY_TABLE) + " WHERE accountId = ?)";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            try (ResultSet rs = statement.executeQuery()) {
                if (rs.next()) {
                    return rs.getInt(1);
                }
            }
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to execute channel count for account query", e);
        }
        return 0;
    }

    public Channel getChannelById(String dbId, String categoryId) {
        List<Channel> channels = getAll(" WHERE id=? AND categoryId=?", new String[]{dbId, categoryId});
        return (channels != null && !channels.isEmpty()) ? channels.get(0) : null;
    }

    public Channel getChannelByChannelIdAndAccount(String channelId, String accountId) {
        String sql = "SELECT c.* FROM " + validatedTableName(CHANNEL_TABLE) + " c" +
                " INNER JOIN " + validatedTableName(CATEGORY_TABLE) + " cat ON c.categoryId = cat.id" +
                " WHERE c.channelId = ? AND cat.accountId = ?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, channelId);
            statement.setString(2, accountId);
            try (ResultSet rs = statement.executeQuery()) {
                if (rs.next()) {
                    return populate(rs);
                }
            }
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to execute getChannelByChannelIdAndAccount query", e);
        }
        return null;
    }

    @SuppressWarnings("java:S2077")
    public List<Channel> getChannelsByChannelIdsAndAccount(Collection<String> channelIds, String accountId) {
        if (channelIds == null || channelIds.isEmpty() || accountId == null || accountId.isBlank()) {
            return List.of();
        }

        List<String> effectiveChannelIds = channelIds.stream()
                .filter(id -> id != null && !id.isBlank())
                .distinct()
                .toList();
        if (effectiveChannelIds.isEmpty()) {
            return List.of();
        }

        // The IN-clause shape is derived only from the number of validated ids; values are still bound safely.
        String placeholders = String.join(",", java.util.Collections.nCopies(effectiveChannelIds.size(), "?"));
        String sql = "SELECT c.* FROM " + validatedTableName(CHANNEL_TABLE) + " c" +
                " INNER JOIN " + validatedTableName(CATEGORY_TABLE) + " cat ON c.categoryId = cat.id" +
                " WHERE cat.accountId = ? AND c.channelId IN (" + placeholders + ")";

        List<Channel> channels = new ArrayList<>();
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            for (int i = 0; i < effectiveChannelIds.size(); i++) {
                statement.setString(i + 2, effectiveChannelIds.get(i));
            }
            try (ResultSet rs = statement.executeQuery()) {
                while (rs.next()) {
                    channels.add(populate(rs));
                }
            }
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to execute getChannelsByChannelIdsAndAccount query", e);
        }
        return channels;
    }

    @SuppressWarnings("java:S1141")
    public void saveAll(List<Channel> channels, String dbCategoryId, Account account) {
        Category category = new CategoryDb().getCategoryByDbId(dbCategoryId, account);
        deleteAll(category.getDbId());
        List<Channel> dedupedChannels = dedupeChannelsCaseInsensitive(channels);
        try (Connection conn = connect()) {
            conn.setAutoCommit(false);
            try (PreparedStatement statement = conn.prepareStatement(insertTableSql(CHANNEL_TABLE))) {
                int count = 0;
                for (Channel channel : dedupedChannels) {
                    statement.setString(1, channel.getChannelId());
                    statement.setString(2, category.getDbId());
                    statement.setString(3, channel.getName());
                    statement.setString(4, channel.getNumber());
                    statement.setString(5, channel.getCmd());
                    statement.setString(6, channel.getCmd_1());
                    statement.setString(7, channel.getCmd_2());
                    statement.setString(8, channel.getCmd_3());
                    statement.setString(9, channel.getLogo());
                    statement.setInt(10, channel.getCensored());
                    statement.setInt(11, channel.getStatus());
                    statement.setInt(12, channel.getHd());
                    statement.setString(13, channel.getDrmType());
                    statement.setString(14, channel.getDrmLicenseUrl());
                    statement.setString(15, channel.getClearKeysJson());
                    statement.setString(16, channel.getInputstreamaddon());
                    statement.setString(17, channel.getManifestType());
                    statement.addBatch();
                    if (++count % BATCH_SIZE == 0) {
                        statement.executeBatch();
                    }
                }
                statement.executeBatch();
                conn.commit();
            } catch (SQLException e) {
                conn.rollback();
                throw new DatabaseAccessException("Unable to execute saveAll query", e);
            } finally {
                conn.setAutoCommit(true);
            }
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to connect to database", e);
        }
    }

    private List<Channel> dedupeChannelsCaseInsensitive(List<Channel> channels) {
        if (channels == null || channels.isEmpty()) {
            return List.of();
        }
        Map<String, Channel> uniqueChannels = new LinkedHashMap<>();
        for (Channel channel : channels) {
            if (channel == null) {
                continue;
            }
            uniqueChannels.putIfAbsent(channelComparisonKey(channel), channel);
        }
        return new ArrayList<>(uniqueChannels.values());
    }

    private String channelComparisonKey(Channel channel) {
        String channelId = normalize(channel.getChannelId());
        if (!channelId.isEmpty()) {
            return "id:" + channelId;
        }
        String name = normalize(channel.getName());
        if (!name.isEmpty()) {
            return "name:" + name;
        }
        String cmd = normalize(channel.getCmd());
        if (!cmd.isEmpty()) {
            return "cmd:" + cmd;
        }
        return "fallback:" + System.identityHashCode(channel);
    }

    private String normalize(String value) {
        return value == null ? "" : value.trim().toLowerCase(Locale.ROOT);
    }

    public void deleteByAccount(String accountId) {
        String sql = "DELETE FROM " + CHANNEL_TABLE.getTableName() +
                " WHERE categoryId IN (SELECT id FROM " + CATEGORY_TABLE.getTableName() + " WHERE accountId=?)";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.execute();
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to execute deleteByAccount query", e);
        }
    }

    @Override
    Channel populate(ResultSet resultSet) {
        Channel c = new Channel(
                nullSafeString(resultSet, "channelId"),
                nullSafeString(resultSet, "name"),
                nullSafeString(resultSet, "number"),
                nullSafeString(resultSet, "cmd"),
                nullSafeString(resultSet, "cmd_1"),
                nullSafeString(resultSet, "cmd_2"),
                nullSafeString(resultSet, "cmd_3"),
                nullSafeString(resultSet, "logo"),
                safeInteger(resultSet, "censored"),
                safeInteger(resultSet, "status"),
                safeInteger(resultSet, "hd"),
                nullSafeString(resultSet, "drmType"),
                nullSafeString(resultSet, "drmLicenseUrl"),
                null, // clearKeys map is reconstructed from JSON
                nullSafeString(resultSet, "inputstreamaddon"),
                nullSafeString(resultSet, "manifestType")
        );
        c.setDbId(nullSafeString(resultSet, "id"));
        c.setCategoryId(nullSafeString(resultSet, "categoryId"));
        c.setClearKeysJson(nullSafeString(resultSet, "clearKeysJson"));
        return c;
    }

}

File: server/src/main/kotlin/com/uiptv/db/ChannelDb.kt
package com.uiptv.db

import com.uiptv.model.Account
import com.uiptv.model.Category
import com.uiptv.model.Channel
import java.sql.ResultSet
import java.sql.SQLException
import java.util.LinkedHashMap
import java.util.Locale

object ChannelDb : BaseDb<Channel>(DatabaseUtils.DbTable.CHANNEL_TABLE) {
    private const val BATCH_SIZE = 1000

    @JvmStatic
    fun get(): ChannelDb = this

    @JvmStatic
    fun insert(channel: Channel, category: Category) {
        try {
            SQLConnection.connect().use { conn ->
                conn.prepareStatement(DatabaseUtils.insertTableSql(DatabaseUtils.DbTable.CHANNEL_TABLE)).use { statement ->
                    statement.setString(1, channel.channelId)
                    statement.setString(2, category.dbId)
                    statement.setString(3, channel.name)
                    statement.setString(4, channel.number)
                    statement.setString(5, channel.cmd)
                    statement.setString(6, channel.cmd_1)
                    statement.setString(7, channel.cmd_2)
                    statement.setString(8, channel.cmd_3)
                    statement.setString(9, channel.logo)
                    statement.setInt(10, channel.censored)
                    statement.setInt(11, channel.status)
                    statement.setInt(12, channel.hd)
                    statement.setString(13, channel.drmType)
                    statement.setString(14, channel.drmLicenseUrl)
                    statement.setString(15, channel.clearKeysJson)
                    statement.setString(16, channel.inputstreamaddon)
                    statement.setString(17, channel.manifestType)
                    statement.execute()
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to execute query", e)
        }
    }

    private fun deleteAll(categoryId: String) {
        val sql = "DELETE FROM ${DatabaseUtils.validatedTableName(DatabaseUtils.DbTable.CHANNEL_TABLE)} WHERE categoryId=?"
        try {
            SQLConnection.connect().use { conn ->
                conn.prepareStatement(sql).use { statement ->
                    statement.setString(1, categoryId)
                    statement.execute()
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to execute delete all query", e)
        }
    }

    fun getChannels(dbId: String): List<Channel> = getAll(" WHERE categoryId=?", arrayOf(dbId))

    fun getChannelCountForAccount(accountId: String): Int {
        val sql = "SELECT COUNT(*) FROM ${DatabaseUtils.validatedTableName(DatabaseUtils.DbTable.CHANNEL_TABLE)} " +
            "WHERE categoryId IN (SELECT id FROM ${DatabaseUtils.validatedTableName(DatabaseUtils.DbTable.CATEGORY_TABLE)} WHERE accountId = ?)"
        try {
            SQLConnection.connect().use { conn ->
                conn.prepareStatement(sql).use { statement ->
                    statement.setString(1, accountId)
                    statement.executeQuery().use { rs ->
                        if (rs.next()) {
                            return rs.getInt(1)
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to execute channel count for account query", e)
        }
        return 0
    }

    fun getChannelById(dbId: String, categoryId: String): Channel? =
        getAll(" WHERE id=? AND categoryId=?", arrayOf(dbId, categoryId)).firstOrNull()

    fun getChannelByChannelIdAndAccount(channelId: String, accountId: String): Channel? {
        val sql = "SELECT c.* FROM ${DatabaseUtils.validatedTableName(DatabaseUtils.DbTable.CHANNEL_TABLE)} c" +
            " INNER JOIN ${DatabaseUtils.validatedTableName(DatabaseUtils.DbTable.CATEGORY_TABLE)} cat ON c.categoryId = cat.id" +
            " WHERE c.channelId = ? AND cat.accountId = ?"
        try {
            SQLConnection.connect().use { conn ->
                conn.prepareStatement(sql).use { statement ->
                    statement.setString(1, channelId)
                    statement.setString(2, accountId)
                    statement.executeQuery().use { rs ->
                        if (rs.next()) {
                            return populate(rs)
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to execute getChannelByChannelIdAndAccount query", e)
        }
        return null
    }

    fun getChannelsByChannelIdsAndAccount(channelIds: Collection<String>?, accountId: String?): List<Channel> {
        if (channelIds.isNullOrEmpty() || accountId.isNullOrBlank()) {
            return emptyList()
        }
        val effectiveChannelIds = channelIds.filter { !it.isNullOrBlank() }.distinct()
        if (effectiveChannelIds.isEmpty()) {
            return emptyList()
        }
        val placeholders = List(effectiveChannelIds.size) { "?" }.joinToString(",")
        val sql = "SELECT c.* FROM ${DatabaseUtils.validatedTableName(DatabaseUtils.DbTable.CHANNEL_TABLE)} c" +
            " INNER JOIN ${DatabaseUtils.validatedTableName(DatabaseUtils.DbTable.CATEGORY_TABLE)} cat ON c.categoryId = cat.id" +
            " WHERE cat.accountId = ? AND c.channelId IN ($placeholders)"
        val channels = ArrayList<Channel>()
        try {
            SQLConnection.connect().use { conn ->
                conn.prepareStatement(sql).use { statement ->
                    statement.setString(1, accountId)
                    effectiveChannelIds.forEachIndexed { index, id -> statement.setString(index + 2, id) }
                    statement.executeQuery().use { rs ->
                        while (rs.next()) {
                            channels += populate(rs)
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to execute getChannelsByChannelIdsAndAccount query", e)
        }
        return channels
    }

    fun saveAll(channels: List<Channel>, dbCategoryId: String, account: Account) {
        val category = requireNotNull(CategoryDb().getCategoryByDbId(dbCategoryId, account)) {
            "Category not found for dbCategoryId=$dbCategoryId"
        }
        deleteAll(category.dbId ?: "")
        val dedupedChannels = dedupeChannelsCaseInsensitive(channels)
        try {
            SQLConnection.connect().use { conn ->
                conn.autoCommit = false
                try {
                    conn.prepareStatement(DatabaseUtils.insertTableSql(DatabaseUtils.DbTable.CHANNEL_TABLE)).use { statement ->
                        var count = 0
                        dedupedChannels.forEach { channel ->
                            statement.setString(1, channel.channelId)
                            statement.setString(2, category.dbId)
                            statement.setString(3, channel.name)
                            statement.setString(4, channel.number)
                            statement.setString(5, channel.cmd)
                            statement.setString(6, channel.cmd_1)
                            statement.setString(7, channel.cmd_2)
                            statement.setString(8, channel.cmd_3)
                            statement.setString(9, channel.logo)
                            statement.setInt(10, channel.censored)
                            statement.setInt(11, channel.status)
                            statement.setInt(12, channel.hd)
                            statement.setString(13, channel.drmType)
                            statement.setString(14, channel.drmLicenseUrl)
                            statement.setString(15, channel.clearKeysJson)
                            statement.setString(16, channel.inputstreamaddon)
                            statement.setString(17, channel.manifestType)
                            statement.addBatch()
                            count++
                            if (count % BATCH_SIZE == 0) {
                                statement.executeBatch()
                            }
                        }
                        statement.executeBatch()
                    }
                    conn.commit()
                } catch (e: SQLException) {
                    conn.rollback()
                    throw DatabaseAccessException("Unable to execute saveAll query", e)
                } finally {
                    conn.autoCommit = true
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to connect to database", e)
        }
    }

    fun deleteByAccount(accountId: String) {
        val sql = "DELETE FROM ${DatabaseUtils.DbTable.CHANNEL_TABLE.tableName} " +
            "WHERE categoryId IN (SELECT id FROM ${DatabaseUtils.DbTable.CATEGORY_TABLE.tableName} WHERE accountId=?)"
        try {
            SQLConnection.connect().use { conn ->
                conn.prepareStatement(sql).use { statement ->
                    statement.setString(1, accountId)
                    statement.execute()
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to execute deleteByAccount query", e)
        }
    }

    override fun populate(resultSet: ResultSet): Channel {
        val channel = Channel(
            nullSafeString(resultSet, "channelId"),
            nullSafeString(resultSet, "name"),
            nullSafeString(resultSet, "number"),
            nullSafeString(resultSet, "cmd"),
            nullSafeString(resultSet, "cmd_1"),
            nullSafeString(resultSet, "cmd_2"),
            nullSafeString(resultSet, "cmd_3"),
            nullSafeString(resultSet, "logo"),
            safeInteger(resultSet, "censored"),
            safeInteger(resultSet, "status"),
            safeInteger(resultSet, "hd"),
            nullSafeString(resultSet, "drmType"),
            nullSafeString(resultSet, "drmLicenseUrl"),
            null,
            nullSafeString(resultSet, "inputstreamaddon"),
            nullSafeString(resultSet, "manifestType")
        )
        channel.dbId = nullSafeString(resultSet, "id")
        channel.categoryId = nullSafeString(resultSet, "categoryId")
        channel.clearKeysJson = nullSafeString(resultSet, "clearKeysJson")
        return channel
    }

    private fun dedupeChannelsCaseInsensitive(channels: List<Channel>?): List<Channel> {
        if (channels.isNullOrEmpty()) {
            return emptyList()
        }
        val uniqueChannels = LinkedHashMap<String, Channel>()
        channels.filterNotNull().forEach { channel ->
            uniqueChannels.putIfAbsent(channelComparisonKey(channel), channel)
        }
        return ArrayList(uniqueChannels.values)
    }

    private fun channelComparisonKey(channel: Channel): String {
        val channelId = normalize(channel.channelId)
        if (channelId.isNotEmpty()) return "id:$channelId"
        val name = normalize(channel.name)
        if (name.isNotEmpty()) return "name:$name"
        val cmd = normalize(channel.cmd)
        if (cmd.isNotEmpty()) return "cmd:$cmd"
        return "fallback:${System.identityHashCode(channel)}"
    }

    private fun normalize(value: String?): String = value?.trim()?.lowercase(Locale.ROOT).orEmpty()
}
